src/org/opensolaris/opengrok/configuration/messages/Message.java

```diff
@@ -282,113 +188,214 @@ public boolean equals(Object obj) {
                         String.format(
                                 "Message was not accepted by the remote server %s%s%s(error %#x).",
                                 message.length() > 0 ? "- " : "",
                                 message.length() > 0 ? message : "",
                                 message.length() > 0 ? " " : "",
-                                c));
+                                deliveryStatus.statusCode));
+        }
+    }
+
+    public enum DeliveryStatus {
+
+        OK(0x1), OVER_LIMIT(0x2), ERROR(0x4);
+
+        private final int statusCode;
+
+        DeliveryStatus(final int statusCode) {
+            this.statusCode = statusCode;
+        }
+
+        public static DeliveryStatus fromStatusCode(final int statusCode) {
+            for (DeliveryStatus ds : values()) {
+                if (ds.statusCode == statusCode) {
+                    return ds;
+                }
+            }
+            throw new IllegalArgumentException("Unknown status code " + statusCode);
+        }
+
+        public int getStatusCode() {
+            return statusCode;
+        }
+    }
+
+    private static class MessageDeserializer implements JsonDeserializer<Message> {
+
+        @Override
+        public Message deserialize(
+                final JsonElement jsonElement,
+                final Type type,
+                final JsonDeserializationContext context
+        ) {
+            String classStr = jsonElement.getAsJsonObject().get("type").getAsString();
+            Class<?> cl;
+            try {
+                cl = Class.forName(classStr);
+            } catch (ClassNotFoundException e) {
+                throw new JsonParseException(e);
+            }
+            return context.deserialize(jsonElement, cl);
+        }
+    }
+
+    private static class DurationAdapter extends TypeAdapter<Duration> {
+
+        @Override
+        public void write(final JsonWriter writer, final Duration duration) throws IOException {
+            if (duration == null) {
+                writer.nullValue();
+                return;
+            }
+            writer.value(duration.toString());
+        }
+
+        @Override
+        public Duration read(final JsonReader reader) throws IOException {
+            if (reader.peek() == JsonToken.NULL) {
+                reader.nextNull();
+                return null;
+            }
+            return Duration.parse(reader.nextString());
+        }
+    }
+
+    public static class Builder<T extends Message> {
+
+        private final Class<T> messageClass;
 
+        private final Set<String> tags = new TreeSet<>();
+
+        private String text;
+
+        private String cssClass;
+
+        private Duration duration;
+
+        public Builder(final Class<T> cl) {
+            messageClass = cl;
+        }
+
+        public T build() {
+            T result = createMessage(messageClass);
+
+            ((Message) result).tags.addAll(tags);
```

Comment:
can you avoid the casting so many times?

Reply:
I can use another variable (of `Message` type -> I actually had it like that at first) but that seemed even worse because it was hiding the intent. Also I would like to avoid casting to `T` since that would cause a warning.

Reply:
And this won't work? I see that it kinda tweaks the approach that one changes some variable which isn't returned at the end of the function but it avoids the casting which is not necessary here to perform it 10 times.

```java
T result = createMessage(messageClass);
Message resultMessage = ((Message) result);

resultMessage.tags.addAll(tags);

if (resultMessage.tags.isEmpty()) {
    Set<String> defaultTags = result.getDefaultTags();
    if (defaultTags != null) {
        resultMessage.tags.addAll(defaultTags);
    }
}

if (text != null) {
    resultMessage.text = text;
}

if (cssClass != null) {
    resultMessage.cssClass = cssClass;
} else {
    resultMessage.cssClass = result.getDefaultCssClass();
}
if (duration != null) {
    resultMessage.duration = duration;
}
return result;
```

Reply:
This is how I had it before but I didn't like it that much because it is weird a little bit :D if you look at the end it seems weird that you are returning `result` but updating `resultMessage`

Not sure what is better though...

Maybe someone else could decide? :D

Reply:
@lkosco

Reply:
Or maybe you can introduce a method with parameter Message and fill the fields inside and call this method here and it does not seem that awful suddenly :-D

Reply:
Yeah, that will work :D